Java build system enhancements

yuzawa-san opened the pull request:
**Description**: I cleaned up the CMake files to delegate building and testing to a Gradle build system. Gradle is manages java build/testing, dependencies, and packaging better than CMake. Additionally, all of the Java build output files are in a single output directory. I added a README which explains what JAR files and build output is generated. I relaxed the requirements in JNI loader in OnnxRuntime.java to allow for flexible loading of the library (i.e. from the os libpath or if the developer wants to load the library from some other way). If the developer wants to do that, then they would not use the `onnxruntime-lib.jar`. I attempted to build out an azure pipeline for the java build (based on the python build). I do not know how to test that. I would assume the maintainers would have to add it. I added spotless plugin to ensure formatting compliance. I selected a stock formatting profile (google's) and reformatted the project in a consistent manner. All changes to java files are formatting except OnnxRuntime.java. Not in this PR: I am playing around with Gradle's maven publishing mechanism.

**Motivation and Context**
- Why is this change required? What problem does it solve?
> Per my original issue below, I wanted to improve the Java packaging and maintainability. I feel these changes will eventually make way for proper maven distribution.
- If it fixes an open issue, please link to the issue here.
> This is relevant towards https://github.com/microsoft/onnxruntime/issues/2675

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/microsoft/onnxruntime?pullRequest=2866) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/onnxruntime?pullRequest=2866) before we can accept your contribution.<br/><br/>:x: yuzawa-san [sign now](https://cla.opensource.microsoft.com/microsoft/onnxruntime?pullRequest=2866)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/microsoft/onnxruntime?pullRequest=2866) it.</sub>

snnn commented:
I'd prefer to remove all the Java related things out of cmake. We don't have to use cmake driving the java compilation.  In order to make it part of the CI, we can add one extra step in win-ci-pipeline.yml. And we may directly use maven/gradle to do so. All of our Windows build machines already have gradle installed at %GRADLE_HOME%. You may invoke it directly.

yuzawa-san commented:
Yeah I moved all of the Java stuff out of the CMake except the piece that compiles the JNI shared object, which needs to stay. I chose the Gradle wrapper for ease of developers, since they won't have to install it and we can lock the version for compatibility purposes. The normal CI process should work just fine since the `--build_java` flag is in there already, so I guess we'll see shortly once this runs. The pipeline I added was a new one for packaging and it creates an azure artifacts `onnxruntime4j` and `onnxruntime4j_gpu`. My vision is that these resulting artifacts can be published to maven for versioned releases and for nightly snapshot releases, just like the python stuff is pushed to PyPi: https://dev.azure.com/onnxruntime/onnxruntime/_release?_a=releases&view=mine&definitionId=1 Regardless how we want to build with respect to CI pipelines, the output is cleaner and all in `build/$OS/$CONFIG/java/build`.

snnn commented on java/src/main/native/ai_onnxruntime_OnnxTensor.h:
```diff
@@ -0,0 +1,133 @@
+/* DO NOT EDIT THIS FILE - it is machine generated */
```
If it's generated, can it be generated on the fly?

yuzawa-san replied:
These files were originally generated in https://github.com/microsoft/onnxruntime/pull/2866/files#diff-81d4ae4b5ee78c2401305edd9d2bcde6L44

However, I took the java stuff out of cmake. The gradle compileJava task generates these files now: https://github.com/microsoft/onnxruntime/pull/2866/files#diff-9d225bc39d80f8e6874619bca156dfc7R39

I attempted to have CMake call gradlew compileJava to generate the headers then compile the onnxruntime4_jni target and then run the gradle cmakeBuild task upon after that is completed, but I had trouble getting onnxruntime4_jni to depend on the gradlew compileJava task. I'll take a look at trying this again shortly.

Craigacp replied:
This is why the build moved entirely to CMake as orchestrating the CMake & gradle dance necessary to have it emit the headers on a clean checkout was tricky. It basically always failed the first time through, such that I had to run gradle skipping tests and then rerun it.

yuzawa-san commented:
The gradle wrapper jar needs to be there since it is a part of the wrapper mechanism https://docs.gradle.org/current/userguide/gradle_wrapper.html#sec:adding_wrapper see the note in that section

snnn commented:
> The gradle wrapper jar needs to be there since it is a part of the wrapper mechanism https://docs.gradle.org/current/userguide/gradle_wrapper.html#sec:adding_wrapper see the note in that section

It's auto-generated right? Then why not generated it in the build?

snnn commented:
Can we avoid using Gradle Wrapper? I'd prefer to not download something from a third-party website during the build, because it adds extra risk.

yuzawa-san commented:
OK I have taken this round of comments into consideration:
* I will use the system gradle installation, falling back to the gradle wrapper. I am hesitant to remove the gradle wrapper fully, since for developers it is a good way to lock the gradle version.
* I was able to generate the header files properly. I basically faked the 'add_jar' CMake target by replacing that with a custom target which calls to gradle to generate a jar. That jar is a pseudo dependency to the onnxruntime4j_jni target. I tested that any alterations to the java or c jni will trigger a rebuild. It appears to work
* I added a call to System.load to attempt to load from the lib path. Logging is included as well. If it fails, then it tries the resources. @Craigacp  I believe this should fix your issue?
* I punted the spotless formatting changes for clarity. It is commented out and I will open another PR at a later date.

yuzawa-san commented:
I also found that gradle wrapper can do checksum check to ensure the wrapper is legitimate https://docs.gradle.org/current/userguide/gradle_wrapper.html#configuring_checksum_verification

Craigacp commented on java/src/main/java/ai/onnxruntime/OnnxRuntime.java:
```diff
@@ -5,175 +5,91 @@
+  private static void load(String library) throws IOException {
+    try {
+      logger.log(Level.FINE, "Attempting to load native library '" + library + "' from lib path");
+      System.load(library);
```
This needs to be a `System.loadLibrary` call, as the library name isn't a fully qualified path.

Craigacp commented:
I think that approach should work once you've switched the method you're calling. What Java versions have you tested this with? I was using 8 and 13 on my main dev machines, and I'm pretty sure I ran it through 11 once or twice to check. `InferenceTest.java` still has the formatting changes applied, so if you could move that to the formatting PR it would be helpful.

Craigacp commented:
I recommend extensive testing of this as it's easy to have CMake ignore the Java bits and not rebuild the jars, headers or binaries as appropriate. The CMake build file that's in the repo doesn't have any oddities in it that I'm aware of, but while I was working on it there were several versions where it looked like everything worked but it hadn't updated one of the jar, the JNI headers or the binaries.

snnn commented on java/src/main/java/ai/onnxruntime/OnnxRuntime.java:
```diff
@@ -41,16 +41,16 @@ static synchronized void init() throws IOException {
    * @param library The bare name of the library.
    * @throws IOException If the file failed to read or write.
    */
   private static void load(String library) throws IOException {
```
Is the argument a full path or relative path?

yuzawa-san replied:
The argument is the library name. System.loadLibrary looks at the configured lib path and it takes the library name. If that fails, then the mappedLibraryName is used to find a file in the java resources.

Craigacp replied:
Might be worth noting that this method calls `System.loadLibrary` in the first case, as you explicitly mention `System.load` for the latter.

snnn replied:
Just one question, what if there are multiple "onnxruntime.dll" in the the configured lib path? For example, if someone put a "onnxruntime.dll" in the "C:\Windows\System32" folder, what's the behavior?

Craigacp replied:
It's implementation dependent (i.e. JVM implementation dependent). I think OpenJDK will iterate the `java.library.path` and load the first one it finds, but I'm not sure.

snnn replied:
It's important to us to make it clear.  For example, what's the default value of "java.library.path" and what's the default behavioral if the user already have an "onnxruntime.dll" in "C:\Windows\System32", is it possible for him to use the file in the jar instead?

yuzawa-san replied:
I was about to update the docs to make it clearer, but I realized that I think I got the load order backwards. It should load from the classpath using the temp file and System.load() and _then_ fall back to the java.library.path and System.loadLibrary(). This ordering is so the specifically included one in the JAR on the classpath is loaded first before looking at the system. So it is more specific to more general. I could imagine somebody putting a version of onnxruntime on the lib path for one use case and then later using the Java API (choosing to use the library included in the JAR) and in that case it should look in the more specific place (the JAR) prior to looking in the more general place (system lib path). Thoughts? It is a simple change that seems to make logical sense. And of course I will update the javadoc regardless of what we decide.

Craigacp replied:
Loading from the classpath first doesn't allow me to use an immutable container unless you're very careful about catching the exceptions as the classpath load will fail with some kind of IOException. The old code did these things, if you supplied a runtime argument to the JVM then it loaded from java.library.path, otherwise by default it loaded from the classpath.

yuzawa-san replied:
I removed the fallback mechanism. In my experimental branch, the logic is now if-else: https://github.com/yuzawa-san/onnxruntime/blob/java-gradle-loader/java/src/main/java/ai/onnxruntime/OnnxRuntime.java#L56
What do y'all think of this? If you wanted to load from lib path, then you would omit the jars with the shared libraries. And there is no fallback. It is disjoint based on the presence of the resources. To use the lib path, you would not add the jar with the libraries to the class path.

Craigacp replied:
I'm a little wary of removing all control as users don't tend to be as in control of their classpath as we might like to think they are. If the onnx runtime is 3 nodes deep in their dependency graph (as it is in some projects I have), then the user relies upon two other library developers making the same decisions about including the native libraries jar or not, and they have no way of controlling that (especially if the jars are shaded).

However, I think that if you do it this way then you should catch the `UnsatisfiedLinkException` that comes out of this method and rethrow it with a nicer message that tells the user what needs to happen (i.e. they need to add the libraries jar to their classpath, or they need to add the libraries to `java.library.path`).

yuzawa-san replied:
I made the loader much more flexible by looking at properties. I'm hoping this would solve any issues that users may have. One property allows for the user to load their own libraries. Another property allows the user to specify an explicit filesystem path to the shared. library file. In the absence of the flags, then the logic checks the resources. If the resource is present then it is used otherwise the System.loadLibrary is used.

Craigacp replied:
I'd make the library names public static final constants on this class, as they are now part of the API. Also I'm unclear why this class became `public` instead of package private as it was before. If it's to make the Javadoc more visible then I recommend putting that javadoc in the `package-info.java` for `ai.onnxruntime`.

yuzawa-san replied:
done. I originally did make it public for the javadoc only.

yuzawa-san commented:
Fixed the loader. I have tested in java 8, but will plan to test in a few later versions. I have manually tested the CMake changes and it looked successful with file changes/moved/adds. I made it so that alterations to the gradle files also cause a rebuild. I have reverted the additional formatting changes.

Craigacp commented on java/src/main/java/ai/onnxruntime/OnnxRuntime.java:
```diff
@@ -5,175 +5,91 @@
+      System.load(temp.getAbsolutePath());
+      logger.log(Level.FINE, "Loaded native library '" + library + "' from resources");
+    } finally {
+      temp.delete();
```
This probably should be `temp.deleteOnExit()`, I'm not sure that the delete trigger will take effect if the library has been loaded.

snnn replied:
If the temp is writable by the other users, it's kind of a security hole.

For example, it could be:
1. A privileged  user got the onnxruntime.jar and started to run it.
2. A temp file was created under the temp directory
3. A hacker found the file and replaced it with something else
4. The dll(or so) was loaded by the privileged user and run.
5. The hacker gain unauthorized access.

yuzawa-san replied:
The internet is pretty undecided about the proper way to load shared libraries from the classpath. I've worked with Netty's native library loader in the past and they seem to employ the very same tempfile method. I don't know if that is best practice, but that's been around for a while. They also by default do a delete like I did. There is a deleteOnExit option, but I do not know why that use case exists. Do you have a deleteOnExit use case? Once the System.load call is done, the library should be in virtual memory and the file can be removed fully. https://github.com/netty/netty/blob/4.1/common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java#L234

In regards to security in the scenario given, I think timing (the short amount of time the file exists) and the random number in the temp filename that Java adds may shield us for the most part. I suppose a hacker could try to predict the random number and somehow maybe listen to file changes and just perfectly swap the file right at just the right moment. a quick internet search of `java file.createTempFile security` did not turn up much except for advising to not create directories.

Craigacp replied:
As to the security issue, I think by default the file on Linux and macOS should only be writable by the user who created it, so I'm not too worried about that. If there is a privileged attacker on the system who can overwrite users' files then there are bigger problems. A quick check with jshell on Windows should determine what default permissions are applied to the temp file.

You can control the permissions that are used to create the file via [this method](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#createTempFile-java.nio.file.Path-java.lang.String-java.lang.String-java.nio.file.attribute.FileAttribute...-), but I'm not sure if the allowable attributes differ across Windows, macOS and Linux.

I think some OSs don't allow you to delete a file that has been dlopened until it's no longer in use. Windows has this issue, it won't let you delete a dll that's in use until the process has gone away (though actually on Windows that means even the `deleteOnExit` hook isn't enough).

yuzawa-san replied:
I do not have a windows runtime available to me to test this out, but I took the netty loader as inspiration and will deleteOnExit if the delete fails (presumably because the file is still open). It should fail in the CI job I believe if this is not the right solution.

Craigacp replied:
I don't think it will fail in the CI, as the deleteOnExit call isn't going to fail then, and when the shutdown hook happens the JVM is going away anyway so I think it gets swallowed. But this isn't too much of an issue as deleteOnExit is the best you can do in the JDK.

Craigacp commented on .gitignore:
```diff
@@ -41,6 +41,5 @@ onnxprofile_profile_test_*.json
 /csharp/src/Microsoft.ML.OnnxRuntime/Microsoft.ML.OnnxRuntime.props
 cmake/external/FeaturizersLibrary/
 # Java specific ignores
-java/src/main/native/ai_onnxruntime_*.h
```
This ignore line should come back now the JNI headers aren't checked in.

yuzawa-san replied:
I don't think it is needed, since the headers are generated in the java/build/headers directory and which is already ignored.

Craigacp replied:
Ok.

Craigacp replied:
The CMake file still looks for them in `java/src/main/native` on line 42?

yuzawa-san replied:
Line 42 refers to any manually created files like `OrtJniUtil.h`. If somebody wanted to add another util or helper object then it would automatically be picked up by CMake and git.

The generated headers are referred to at the end of line 51:
```
target_include_directories(onnxruntime4j_jni PRIVATE ${REPO_ROOT}/include ${JAVA_ROOT}/build/headers)
```

yuzawa-san commented:
@snnn how do we test the azure pipeline? I assume you have to install it? Should it be a separate PR?

snnn commented:
@yuzawa-san  could you please remove the jar from your changes? I'd prefer to not checkin any executable binary in git.

yuzawa-san commented:
So the users will need to bring their own gradle instead of the wrapper? what about configuring verification https://docs.gradle.org/current/userguide/gradle_wrapper.html#sec:verification to guarantee the integrity?

snnn commented:
> So the users will need to bring their own gradle instead of the wrapper?

I feel it's good.

yuzawa-san commented:
@snnn ok I removed the gradle  wrapper, but gitignored any wrappers in the event developers want to use their own wrapper (so they won't get committed). I moved the azure pipeline stuff to another branch to be used in future PR.

@Craigacp I think I am done. I'm going to test in a few more environments. Does it look ok in your environments?

snnn commented:
/azp run Linux CPU CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline,MacOS CI Pipeline,Win CPU x86 CI Pipeline,Windows CPU CI Pipeline,Windows GPU CI Pipeline,Windows GPU TensorRT CI Pipeline

yuzawa-san commented:
@snnn the logs show that the Linux CI is missing gradle. Alternatively, gradle is not on the PATH

snnn commented:
Does gradle 2.10 meet your need?

yuzawa-san commented:
Ideally, we would use Gradle version 6 or newer. Gradle 2 is from 2014 and probably way out of date. It may work with Gradle 4 or 5 which are from the last few years, but I cannot guarantee that will work.

yuzawa-san commented:
I'm looking at the logs for the windows failures. I can't really see why it was taking long. It almost looks like it was stuck. Any ideas? Unfortunately, I do not have access to any windows environments to reproduce.

Craigacp commented on java/src/main/java/ai/onnxruntime/OnnxRuntime.java:
```diff
@@ -5,175 +5,116 @@
+      logger.log(
+          Level.FINE,
+          "Attempting to load native library '" + library + "' from resource path " + resourcePath);
+      File temp = File.createTempFile("javaload_", libraryFileName);
```
Looking at #2892 this might cause an issue on Windows. I'll try to test out my proposed fix this week if I can persuade my desktop to have the right Windows build environment (once I've checked it doesn't change macOS or Linux), and then we'd need to incorporate that change into this loader.

yuzawa-san replied:
OK sounds good, since I do not have a windows environment.

Do we think this is the reason the windows CI failed? @snnn should we rerun that part?

Taking a look at the netty loader: 
https://github.com/netty/netty/blob/4.1/common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java#L183

They set the prefix and suffix such that `onnxruntime.dll` becomes `onnxruntime123452.dll` maybe the prefix needs to match like that for windows to play nicely. However your original work seems to do this already. Additionally, they have a flush() call. Maybe that is something that windows really likes. I have no other ideas at the moment.

Craigacp replied:
I have a fix for #2892, https://github.com/Craigacp/onnxruntime/commit/3aef28097a53cdd8bfa9a0a05ed1106a30acef20. Do you want to mirror the logic into this PR? Or I can submit a PR to your branch with the fix.

yuzawa-san replied:
I'll mirror it here shortly

snnn commented:
Sorry, let me finish the other smaller PRs first.

mina-asham commented on java/src/main/java/ai/onnxruntime/OnnxRuntime.java:
```diff
@@ -5,175 +5,114 @@
+    // replace Mac's jnilib extension to dylib
+    String libraryFileName = System.mapLibraryName(library).replace("jnilib", "dylib");
+    String resourcePath = "/ai/onnxruntime/native/" + libraryFileName;
+    InputStream is = OnnxRuntime.class.getResourceAsStream(resourcePath);
```
We should close this at the end

yuzawa-san replied:
done with a try-with-resources

mina-asham commented on java/src/main/java/ai/onnxruntime/OnnxRuntime.java:
```diff
@@ -5,175 +5,114 @@
+      return;
+    }
+    Path tempDirectory = Files.createTempDirectory("onnxruntime-java");
+    tempDirectory.toFile().deleteOnExit();
```
Minor: deleteOnClose is usually a catch-all operation, I think we should try and delete it explicitly as well after loading instead of leaving the files available till the JVM shuts down (same for the load method below)

yuzawa-san replied:
@Craigacp was there some nuance about closing in Windows or some OS that keeps the library open and does not like it being deleted. In an earlier version, I did this 
```java
if(!file.delete()){
 file.deleteOnExit();
}
```
the condition would fail if the OS still has the file open and it would be cleaned up on exit. What do y'all think? Netty's loader does something similar to this: https://github.com/netty/netty/blob/4.1/common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java#L234

Craigacp replied:
I think that would be fine. I tend to just mark things `deleteOnExit` rather than trying to delete it first when I do this in other Java code, but I'm not aware of a specific issue. If Netty do that it should be fine.

yuzawa-san replied:
ok sounds good. I've updated it tested in linux.

yuzawa-san commented:
@snnn any updates on adding gradle (6) to the linux CI?

yuzawa-san commented:
Any progress on the CI fixes?

snnn commented:
Please give me a few days. I'm working it.

mina-asham commented on java/build.gradle:
```diff
@@ -0,0 +1,117 @@
+	//java {
+	//	removeUnusedImports()
+	//	googleJavaFormat()
+	//}
```
Should probably either uncomment or remove?

yuzawa-san replied:
this was originally in this pr and will be in the PR I open immediately after this one is merged

snnn commented:
I need to add a few lines in: https://github.com/microsoft/onnxruntime/blob/master/tools/ci_build/github/linux/docker/scripts/install_deps.sh

To download gradle and unzip it.

yuzawa-san commented:
@snnn to verify, you are updating that file? On this pr or separately?

snnn commented:
/azp run Win CPU x64 NoContribops CI Pipeline,MacOS NoContribops CI Pipeline

snnn commented:
/azp run Linux CPU CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline,MacOS CI Pipeline,Win CPU x86 CI Pipeline,Windows CPU CI Pipeline,Windows GPU CI Pipeline,Windows GPU TensorRT CI Pipeline

snnn commented:
/azp run Linux CPU x64 NoContribops CI Pipeline

snnn commented on java/src/main/java/ai/onnxruntime/OnnxRuntime.java:
```diff
@@ -5,175 +5,135 @@
+            os.write(buffer, 0, readBytes);
+          }
         }
+        System.load(tempFile.getAbsolutePath());
```
So, if onnxruntime.dll has any dependencies that are not in the java.library.path, then this call will fail, right?

yuzawa-san replied:
I do not know precisely since I do not work in windows environments, but I believe the load call will load the file and whatever dependency configuration is set in the library file. Linux SO files can have relative, absolute, and rpath dependency locations. DLL's should probably have similar capabilities.

Craigacp replied:
It should check the system path for dependent libraries as it's basically a wrapper around the platform's dynamic library loader.

yuzawa-san commented:
Gradle still appears to be missing https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=114373&view=logs&j=8c952c3b-dbd4-5685-50d1-7e2a31304cbf&t=e14f62b2-f329-5130-d196-ee9974ccf2a3

snnn commented:
I'm fixing it.

snnn commented:
/azp run Linux CPU CI Pipeline

snnn commented:
@Craigacp Do you think this PR is ready to merge?

Craigacp commented:
@snnn Assuming all the CI pipelines pass, yes.

snnn commented:
/azp run Linux CPU CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline,MacOS CI Pipeline,Win CPU x86 CI Pipeline,Windows CPU CI Pipeline,Windows GPU CI Pipeline,Windows GPU TensorRT CI Pipeline

snnn commented:
/azp run Win CPU x64 NoContribops CI Pipeline,MacOS NoContribops CI Pipeline,Linux CPU x64 NoContribops CI Pipeline

yuzawa-san commented:
Ok cool the linux build works. The windows ci pipelines timed out because they took a long time for unknown reasons. Something is off given that most of the runs for that job take around 25 minutes. Any ideas? Sadly I do not have access to a windows environment, so any help would be appreciated.

Does the gradle daemon keep running or something weird which prevents that stage from ending? I think my last CI build hung indefinitely after those pybind jobs were built. I can see the Java builds occurring in the logs.

snnn commented:
Something is wrong here. Windows build hangs forever.

snnn commented:
Let me try to disable gradle daemon.

snnn commented:
/azp run Linux CPU CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline,MacOS CI Pipeline,Win CPU x86 CI Pipeline,Windows GPU CI Pipeline,Windows GPU TensorRT CI Pipeline

snnn commented:
/azp run Win CPU x64 NoContribops CI Pipeline,MacOS NoContribops CI Pipeline,Linux CPU x64 NoContribops CI Pipeline

snnn commented:
This time it should be ok.
